### DISCUSSION TITLE
Criteria API Extensions / Additions

### DISCUSSION THREAD

**Discussion by JasonTypesCodes** [hibernate/hibernate-orm#4795] [Design Proposals]
Hello Hibernate Team!

The Grails / GORM team is currently investigating ideas to resolve a long-lived issue in GORM related to the deprecation of the Hibernate Criteria API (https://github.com/grails/gorm-hibernate5/issues/55)

The move from the Hibernate Criteria API straight to the JPA API isn't ideal for us due to some functional gaps between the APIs (https://hibernate.atlassian.net/browse/HHH-13170)

As we are working through solutions to this, one idea we had was to create a non-JPA Criteria API extension to your API to fill in some of these gaps and to submit those updates back to Hibernate.  

Before we start down this path, we'd like to know if a change like that would be welcome in the Hibernate code base as it would lead to additional maintenance for your group.  Would you welcome a change of this nature presuming we coordinated on the details of the API in advance?

**Comment by beikov** [comment]
I don't know what you mean by non-JPA Criteria API extension, but if that means to add an additional API, I don't think that this is the right path. If we wanted to maintain multiple query models, we would have kept the legacy Criteria API.

Extension to the JPA Criteria API are very welcome though, given that we discuss these beforehand.

**Comment by sebersole** [comment]
As I think @beikov is alluding to...

```
CriteriaBuilder cb = ...;
// unfortunately no `CriteriaBuilder#unwrap`, so you'll have to cast
HibernateCriteriaBuilder hcb = (HibernateCriteriaBuilder) cb;
```

Any "extensions" would be available from HibernateCriteriaBuilder.  From there, we just need to talk about the individual ones you think are missing.  However, I did scour the old Criteria API to make sure Restrictions, etc that made sense made their way to HibernateCriteriaBuilder.  So we'd need some convincing to add others.

**Comment by JasonTypesCodes** [comment]
Thank you both.  We are going to do some digging on our end.  We'll chime back in here with specifics after doing some research.

**Reply by beikov** [reply]
I'll close this discussion for now. If you have concrete proposals, the please create a new discussion.
